daml-assistant/daml-helper/DamlHelper.hs

```diff
@@ -103,6 +132,58 @@ runListTemplates = do
           "The following templates are available:" :
           map (\dir -> "  " <> takeFileName dir) templates
 
+
+newtype SandboxPort = SandboxPort Int
+newtype NavigatorPort = NavigatorPort Int
+
+sandboxPortReader :: ReadM SandboxPort
+sandboxPortReader = SandboxPort <$> auto
+
+withSandbox :: SandboxPort -> [String] -> (ProcessHandle -> IO a) -> IO a
+withSandbox (SandboxPort port) args a = do
+    withJar sandboxPath (["--port", show port] ++ args) $ \ph -> do
+        putStrLn "Waiting for sandbox to start: "
+        -- TODO We need to figure out what a sane timeout for this step.
+        waitForConnectionOnPort (putStr "." *> threadDelay 500000) port
+        a ph
+
+withNavigator :: SandboxPort -> NavigatorPort -> FilePath -> [String] -> (ProcessHandle-> IO a) -> IO a
+withNavigator (SandboxPort sandboxPort) (NavigatorPort navigatorPort) config args a = do
+    withJar navigatorPath (["server", "-c", config, "localhost", show sandboxPort, "--port", show navigatorPort] <> args) $ \ph -> do
+        putStrLn "Waiting for navigator to start: "
+        -- TODO We need to figure out a sane timeout for this step.
+        waitForHttpServer (putStr "." *> threadDelay 500000) ("http://localhost:" <> show navigatorPort)
+        a ph
+
+runSandbox :: SandboxPort -> [String] -> IO ()
+runSandbox port args = do
+    exitCode <- withSandbox port args waitForProcess
+    exitWith exitCode
+
+runStart :: FilePath -> IO ()
+runStart darPath = do
+    withSandbox sandboxPort [darPath] $ \sandboxPh -> do
+        parties <- getProjectParties
+        withTempDir $ \confDir -> do
+            -- Navigator determines the file format based on the extension so we need a .json file.
+            let navigatorConfPath = confDir </> "navigator-config.json"
+            writeFileUTF8 navigatorConfPath (T.unpack $ navigatorConfig parties)
+            withNavigator sandboxPort navigatorPort navigatorConfPath [] $ \navigatorPh ->
+                void $ race (waitForProcess navigatorPh) (waitForProcess sandboxPh)
+    where sandboxPort = SandboxPort 6865
+          navigatorPort = NavigatorPort 7500
+
+getProjectParties :: IO [T.Text]
+getProjectParties = do
+    projectPath <- required "Must be called from within a project" =<< getProjectPath
```

Comment:
This looks like something that might be useful in multiple places. Should we move it into its own function?

Reply:
I’ll do so when we use it somewhere else.